File: WhatShouldIWorkOnToday/Server/Authentication/BasicAuthenticationHandler.cs
using Microsoft.AspNetCore.Authentication;
using Microsoft.Extensions.Options;
using System.Security.Claims;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.RegularExpressions;

namespace WhatShouldIWorkOnToday.Server.Authentication;

public class BasicAuthenticationHandler : AuthenticationHandler<AuthenticationSchemeOptions>
{
	private readonly IConfiguration _config;

	public BasicAuthenticationHandler(
		IOptionsMonitor<AuthenticationSchemeOptions> options,
		ILoggerFactory logger,
		UrlEncoder encoder,
		ISystemClock clock,
		IConfiguration config) : base(options, logger, encoder, clock)
	{
		_config = config;
	}

	protected override Task<AuthenticateResult> HandleAuthenticateAsync()
	{
		Response.Headers.Add("WWW-Authenticate", "Basic");

		if(!Request.Headers.ContainsKey("Authorization"))
		{
			return Task.FromResult(AuthenticateResult.Fail("Authorization header missing."));
		}

		var authorizationHeader = Request.Headers["Authorization"].ToString();
		var authHeaderRegex = new Regex(@"Basic (.*)");

		if(!authHeaderRegex.IsMatch(authorizationHeader))
		{
            return Task.FromResult(AuthenticateResult.Fail("Authorization code not formatted properly."));
        }

		var authBase64 = Encoding.UTF8.GetString(Convert.FromBase64String(authHeaderRegex.Replace(authorizationHeader, "$1")));
		var authSplit = authBase64.Split(Convert.ToChar(":"), 2);
		var authUsername = authSplit[0];
		var authPassword = authSplit.Length > 1 ? authSplit[1] : throw new Exception("Unable to get password");

		var validUser = _config["Login"];
		var validPassword = _config["Password"];
		if(authUsername != validUser || authPassword != validPassword)
		{
			return Task.FromResult(AuthenticateResult.Fail("The username or password is not correct."));
		}

		var authenticatedUser = new AuthenticatedUser("BasicAuthentication", true, authUsername);
		var claimsPrinciple = new ClaimsPrincipal(new ClaimsIdentity(authenticatedUser));

		return Task.FromResult(AuthenticateResult.Success(new AuthenticationTicket(claimsPrinciple, Scheme.Name)));
    }
}

File: WhatShouldIWorkOnToday.Service.Host/BasicAuthenticationHandler.fs
namespace WhatShouldIWorkOnToday.Server.Authentication

open Microsoft.AspNetCore.Authentication
open Microsoft.Extensions.Options
open Microsoft.Extensions.Logging
open System.Security.Claims
open System.Text
open System.Text.Encodings.Web
open System.Text.RegularExpressions
open Microsoft.Extensions.Configuration
open System.Threading.Tasks
open System

type BasicAuthenticationHandler (options: IOptionsMonitor<AuthenticationSchemeOptions>, loggerFactory: ILoggerFactory, encoder: UrlEncoder, clock: ISystemClock, config: IConfiguration) =
    inherit AuthenticationHandler<AuthenticationSchemeOptions>(options, loggerFactory, encoder, clock)

    let _config = config

    override __.HandleAuthenticateAsync () =
        base.Response.Headers.Add("WWW-Authenticate", "Basic realm=\"WhatShouldIWorkOnToday\"")

        if not <| base.Request.Headers.ContainsKey("Authorization") then
            Task.FromResult(AuthenticateResult.Fail("Authorization header missing."))
        else
            let authorizationHeader = base.Request.Headers.["Authorization"].ToString()
            let authHeaderRegex = Regex(@"Basic (.*)")

            if not <| authHeaderRegex.IsMatch(authorizationHeader) then
                 Task.FromResult(AuthenticateResult.Fail("Authorization code not formatted properly."))
            else
                let authBase64 = Encoding.UTF8.GetString(Convert.FromBase64String(authHeaderRegex.Replace(authorizationHeader, "$1")))
                let authSplit = authBase64.Split(Convert.ToChar(":"), 2)
                let authUsername = authSplit.[0]
                let authPassword =
                    if authSplit.Length > 1 then
                        authSplit.[1]
                    else
                        failwith "Unable to get password"

                let validUser = "user"          // TODO: Get this from config
                let validPassword = "password"

                if authUsername <> validUser || authPassword <> validPassword then
                     Task.FromResult(AuthenticateResult.Fail("The username or password is not correct."))
                else
                    let authenticatedUser = AuthenticatedUser("BasicAuthentication", true, authUsername)
                    let claimsPrinciple = ClaimsPrincipal(new ClaimsIdentity(authenticatedUser))
                    Task.FromResult(AuthenticateResult.Success(new AuthenticationTicket(claimsPrinciple, base.Scheme.Name)))
